common/scala/src/main/scala/org/apache/openwhisk/core/database/mongodb/MongoDBArtifactStore.scala

```diff
@@ -0,0 +1,661 @@
+              this,
+              start,
+              s"[PUT] '$dbName' failed to put document: '$docinfoStr'; return error code: '${t.getCode}'",
+              ErrorLevel)
+            throw new Exception("Unexpected mongodb server error: " + t.getMessage)
+        }
+
+    reportFailure(
+      f,
+      failure =>
+        transid
+          .failed(this, start, s"[PUT] '$collName' internal error, failure: '${failure.getMessage}'", ErrorLevel))
+  }
+
+  override protected[database] def del(doc: DocInfo)(implicit transid: TransactionId): Future[Boolean] = {
+    require(doc != null && doc.rev.asString != null, "doc revision required for delete")
+
+    val start =
+      transid.started(this, LoggingMarkers.DATABASE_DELETE, s"[DEL] '$collName' deleting document: '$doc'")
+
+    val f = collection
+      .deleteOne(Filters.and(Filters.eq("_id", doc.id.id), Filters.eq("_rev", doc.rev.rev)))
+      .toFuture()
+      .flatMap { result =>
+        if (result.getDeletedCount == 1) { // the result can only be 1 or 0
+          transid.finished(this, start, s"[DEL] '$collName' completed document: '$doc'")
+          Future(true)
+        } else {
+          collection.find(Filters.eq("_id", doc.id.id)).toFuture.map { result =>
+            if (result.size == 1) {
+              // find the document according to _id, conflict
+              transid.finished(this, start, s"[DEL] '$collName', document: '$doc'; conflict.")
+              throw DocumentConflictException("conflict on 'delete'")
+            } else {
+              // doesn't find the document according to _id, not found
+              transid.finished(this, start, s"[DEL] '$collName', document: '$doc'; not found.")
+              throw NoDocumentException(s"$doc not found on 'delete'")
+            }
+          }
+        }
+      }
+      .recover {
+        case t: MongoException =>
+          transid.failed(
+            this,
+            start,
+            s"[DEL] '$collName' failed to delete document: '$doc'; error code: '${t.getCode}'",
+            ErrorLevel)
+          throw new Exception("Unexpected mongodb server error: " + t.getMessage)
+      }
+
+    reportFailure(
+      f,
+      failure =>
+        transid.failed(
+          this,
+          start,
+          s"[DEL] '$collName' internal error, doc: '$doc', failure: '${failure.getMessage}'",
+          ErrorLevel))
+  }
+
+  override protected[database] def get[A <: DocumentAbstraction](doc: DocInfo,
+                                                                 attachmentHandler: Option[(A, Attached) => A] = None)(
+    implicit transid: TransactionId,
+    ma: Manifest[A]): Future[A] = {
+
+    val start = transid.started(this, LoggingMarkers.DATABASE_GET, s"[GET] '$dbName' finding document: '$doc'")
+
+    require(doc != null, "doc undefined")
+
+    val f = collection
+      .find(Filters.eq("_id", doc.id.id)) // method deserialize will check whether the _rev matched
+      .toFuture()
```

Comment:
nit: any thoughts on `.toSingle().toFutureOption()`? Your result would be a `Future[Option[A]]` which looks to match your check for an empty collection or single-item collection below.

Reply:
I'm sorry, I didn't find which object have the `toSingle` method?